### PR TITLE
ERL-453 Optimized away atoms are reintroduced during ASM stage

### PR DESCRIPTION
**Pull request by kvakvs**
The bug https://bugs.erlang.org/browse/ERL-453 says that atoms used in code, which was removed by optimizer, should remain in the module in order for `*_to_existing_atom` to work.

* In sys_core_fold I collect optimized away atoms in special protected ETS table.
* The collected atoms are stored as module attribute `-optimized_away_atoms([a, b, c...]).`, some of atoms here may be used elsewhere in the code and will look like false positives. They are listed here because some piece of code was eliminated which contained such an atom.
* BEAM ASM stage reads this attribute and reintroduces removed atoms into the atom dict.
* The bug is gone.

## Possible problems?
* There is no option to control/disable this feature.
* Possibly a mention in documentation is needed? This does not introduce a new behaviour, it just restores expected behaviour for `*_to_existing_atom` for cases when the atom was optimized away and eliminated.

### CODE REVIEW

**Comment by RaimoNiskanen** [comment]
I'll leave it to the compiler guys to look at this after their vacation, to see if they are happy with your solution.

I do not think this needs an option since an optimization should not change the semantics in any way.

File: lib/compiler/src/beam_asm.erl [hunk 1]
```diff
@@ -68,7 +68,18 @@ assemble({Mod,Exp0,Attr0,Asm0,NumLabels}, ExtraChunks, SourceFile, Opts, Compile
     {Asm,Attr} = on_load(Asm0, Attr0),
     Exp = cerl_sets:from_list(Exp0),
     {Code,Dict2} = assemble_1(Asm, Exp, Dict1, []),
-    build_file(Code, Attr, Dict2, NumLabels, NumFuncs, ExtraChunks, SourceFile, Opts, CompilerOpts).
+
+    %% Populate atom dict with atoms that were optimized away during Core Fold
+    %% Atoms may possibly exist in atom Dict, but we attempt to add them anyway
+    OptAtoms = proplists:get_value(optimized_away_atoms, Attr0, []),
```
Comment (bjorng) [inline lib/compiler/src/beam_asm.erl]: I would prefer a name that can't possible be in conflict with a user-defined attribute. Perhaps you could a binary as a key?

`OptAtoms = proplists:get_value(<<"optimized_away_atoms">>, Attr0, []),

If that would be problematic, perhaps you could add an underscore like this:

`OptAtoms = proplists:get_value(_optimized_away_atoms, Attr0, []),

File: lib/compiler/src/sys_core_fold.erl [hunk 2]
```diff
@@ -113,9 +115,34 @@ module(#c_module{defs=Ds0}=Mod, Opts) ->
 	_ -> ok
     end,
     init_warnings(),
+
+    %% Collect atoms which are optimized away
+    _Tid = ets:new(?OPT_ATOMS_TAB, [named_table, protected]),
```
Comment (bjorng) [inline lib/compiler/src/sys_core_fold.erl]: A named table will prevent running several instances of the compiler at the same time. I suggest that you use a `cerl_sets` that you put in the process dictionary.

File: lib/compiler/src/sys_core_fold.erl [hunk 3]
```diff
@@ -197,6 +224,7 @@ opt_guard_try(#c_seq{arg=Arg,body=Body0}=Seq) ->
 		    %% a guard and that we must keep the call).
 		    Seq#c_seq{body=Body};
 		true ->
+		    optimized_away_atoms([Mod, Name] ++ Args),
```
Comment (bjorng) [inline lib/compiler/src/sys_core_fold.erl]: Write this as:

`optimized_away_atoms([Mod, Name|Args]),`

File: lib/compiler/src/sys_core_fold.erl [hunk 4]
```diff
@@ -786,7 +814,9 @@ fold_call_1(Call, Mod, Name, Args, Sub) ->
     NumArgs = length(Args),
     case erl_bifs:is_pure(Mod, Name, NumArgs) of
 	false -> Call;				%Not pure - keep call.
-	true -> fold_call_2(Call, Mod, Name, Args, Sub)
+	true ->
+	    optimized_away_atoms([Mod, Name] ++ Args),
```
Comment (bjorng) [inline lib/compiler/src/sys_core_fold.erl]: Write this as:

`optimized_away_atoms([Mod, Name|Args]),`

**Comment by kvakvs** [comment]
I have a test for this, if we compile the compiler app with itself, then it might work. Adding it as a separate commit soon.

**Comment by kvakvs** [comment]
Can't get it to fail on master, though :(

**Comment by bjorng** [comment]
> Can't get it to fail on master, though :( It is placed in a SUITE, and does not fail. If I run a similar test by compiling from shell and then loading the module — it works (i.e. fails and shows the problem). Are suites themselves compiled without this optimization? Should i use compile:file and a separate file?

I can't fully explain why the test case doesn't fail. I can explain why it doesn't fail the first time you run the test suite -- the test suite will be compiled when running the tests suite and the compiler will create the atom. However, I would expect that it would fail if I run it a second time, but it doesn't.

I suggest that you use a separate module as you suggested, and that you use a slave node to run load and run the compiled code. (There are examples on how to use slave nodes in `inline_SUITE.erl`.)

**Comment by kvakvs** [comment]
I'll try and produce a test today.

**Comment by kvakvs** [comment]
I implemented the test without starting a separate node. I write test file with a random atom, do `file:compile` on it and check if the atoms section of BEAM contains my random atom.

**Comment by bjorng** [comment]
I get 5 failed test cases in the compiler test suite, and your test case still does not fail if I remove the fix in the compiler.

I have been thinking whether we should do this at all, and I have the reached the conclusion that it should not be the responsibility of the compiler to avoid losing optimized-away atoms. There are too many ways atoms can be optimized away (e.g. if inlining is turned on). The responsibility should be on the user of list_to_existing_atom/1. (Also see my comment in https://bugs.erlang.org/browse/ERL-453.)

**Pull request closed without merging**